### ISSUE TITLE
Meteor for Windows unstable

### ISSUE THREAD

**Issue by AlbinoGeek** [meteor/meteor#5571]
Hello all,

After having used Meteor for Windows extensively, I would like to report for sure that this is an issue in the hot-code-push that is used on Windows (the rebuilding and re-linking that didn't exist prior to version 1.2, but is a huge problem now.)

This problem causes the system PID to sky-rocket (past 700,000), and on Microsoft Windows; the PID can never decrease.  This, among other issues (such as leaking handles); causes the system to become completely unusable after approximately 100 hot-code-pushes (i.e: file changes while the server is running.)

Even 100 "Client Refreshes" can cause this issue, it is not necessarily related to the server rebuilding.  The rate of system degradation has to do with your maximum RAM, as I have 16GB I can generally "put up with" this issue for some 8 hours before I have to reboot, every day, multiple times a day.

This issue is not noticeable in _nix, while it *_may exist**, the garbage collection in the kernel deals with it.  Microsoft Windows has no garbage collection for PIDs or handles, therefore this issue is huge on Windows 10, and 8.1, and 8.  I have not tested this issue on Windows 7 yet.

---

Symptoms visible of degradation:
- All new processes have an extremely high (600,000+) PID
- Services start leaking resources (virtual memory)
- Many orphaned handles are left in the system (visible via handles*)

I have isolated this issue in Oracle Virtualbox to see that it was not hardware related.

---

As per how to fix this, I'm not sure.  This probably requires more research and some memory dumps to work with, I will start on providing any information as requested.  I am currently running my tests from Windows 10 Professional x64 inside a Oracle Virtualbox latest (5.x), with a host system of CentOS Linux 7.

---

I've avoided posting this for some time, as I wasn't able to reproduce the issue in a short amount of time (it is a build up issue, you must be developing / using the system to cause the issue to happen / to irritate the issue.)  One thing you can do to see the issue aggravate itself: open `Administrator Task Manager` and watch the Performance tab, Memory graph.  Also `Administrator Resource Monitor` will show you the leaking handles and ever-increasing PID.

**Comment by AlbinoGeek** [comment]
Similar to #5569 the code pushes do take longer and longer as the problem gets worse.

**Comment by AlbinoGeek** [comment]
I hope people are not being spammed from this thread, but I will update it every so often with more information; memory dumps; process explorer maps; and screenshots as the problem persists, this has made Meteor very difficult for us to use.

https://db.tt/2bh4HTz7

I had called taskkill 30 seconds after starting the `meteor` command in the project directory.  I then waited 30 seconds before taking the screenshot (see the large gap in PIDs.)

Notice in this screenshot the large gaps between process IDs (caused by the superspawning of builder and link processes, as any compilation normally would (perhaps we could reduce these using exec?  ... nevermind, Microsoft Windows has no exec.)  One thing you can do is "call link only once", but yeah... still won't help the compiler PIDs.

---

If I take a screenshot where you can see all the dead processes 5 minutes later:

Another jump in PIDs: https://db.tt/cYmMplA7

And if I force a hot-code-refresh by adding (, waiting 10 seconds, then removing) a newline in a file...

More large PID gaps, and increased RAM usage: https://db.tt/GeNKu7ll

**Comment by AlbinoGeek** [comment]
This is where the system starts getting unusable:

##### https://db.tt/qegm1Fss (some hours later), PID is now past 215,000

**Comment by AlbinoGeek** [comment]
#### Found a temporary work-around.

If you want to delay your required restart, you can **aggressively reclaim your memory** using Microsoft Sysinternals RAMMap v1.14 (released October 26th, 2015 to include support for Microsoft Windows 10.)

(Microsoft Technet - Sysinternals RAMMap)[https://technet.microsoft.com/en-us/library/ff700229.aspx]

-1 Extract rammap.exe from downloaded Zip.
-2 Run as Administrator.
-3 If you receive an error, press OK and then F5
-4 Repeat step 3 until you get a memory graph (takes 1-4 tries)
-5 With a graph, go to the "Empty" menu, and start from the bottom.

My case is extreme to the point where I need to dump all `Working Sets` (every menu item, from bottom to top) to force anything to be freed.  The system stutters for ~2ms, and Task Manager shows a rather aggressive write to disk as the entire "Modified" and "Standby" are emptied.

The PIDs still climb, hot-code-reloads are still slow, but you get ~6GB of RAM back each time.

**Comment by AlbinoGeek** [comment]
##### `86%` RAM Used , `1,077,896` PIDs later:

https://db.tt/fu0j7hQ8

And after killing everything with fire (rammap) [my workaround]:

##### still 69% memory in use, 1million PIDs can't be fixed without reboot.

https://db.tt/cG4CoVzV

**Comment by AlbinoGeek** [comment]
However; everything's fine again after a reboot:

https://db.tt/iA07D4PF

But not once I start the empty application just once again:

https://db.tt/OIQErVns

**Comment by stubailo** [comment]
Thanks for the in-depth analysis - is there something in particular we could do to alleviate this problem?

**Comment by AlbinoGeek** [comment]
@stubailo I have found a similar issue today with "Git for Windows" found her:

- http://stackoverflow.com/a/18503369 (workaround rebase shared memory)
- http://stackoverflow.com/a/10114597 (rebase.exe renamed in Windows SDK later...)

Using `editbin /REABSE` I was able to alleviate some of my memory problems.  I'll put together a script in the next few days of the exact object names that have to be rebased, and where they can safely be rebased to.

The only other workaround that is still working is using `rammap` and dumping absolutely everything (including the `System Working Sets` and `Working Sets`; which probably affects other programs and system services as a side-effect.)

**Comment by hwillson** [comment]
Several versions of Meteor have been released since this issue was created, that include many bug fixes and performance improvements. Given the lack of activity on this issue in over a year, and the fact that many developers are successfully developing on Windows with the current version of Meteor, I'll close if off. If you notice any issues with the latest Meteor version, please open a new issue. Thanks for reporting these in-depth findings originally!
